kale/evaluate/uncertainty_metrics.py

```diff
@@ -0,0 +1,1036 @@
+
+            fold_uncertainty_values = np.array(fold_uncertainty_values.values.tolist()).flatten()
+
+            # Get the quantiles of the data w.r.t to the uncertainty values.
+            # Get the true uncertianty quantiles
+            sorted_uncertainties = np.sort(fold_uncertainty_values)
+
+            quantiles = np.arange(1 / num_bins_quantiles, 1, 1 / num_bins_quantiles)[: num_bins_quantiles - 1]
+            quantile_thresholds = [np.quantile(sorted_uncertainties, q) for q in quantiles]
+
+            # If we are combining the middle bins, combine the middle lists into 1 list.
+            if combine_middle_bins:
+                quantile_thresholds = [quantile_thresholds[0], quantile_thresholds[-1]]
+
+            # fit a piece-wise linear regression line between quantiles, and return correlation values.
+            # os.makedirs(os.path.join(save_path, model, uncertainty_type), exist_ok=True)
+
+            save_path_fig = (
+                os.path.join(save_path, model + "_" + uncertainty_type + "_correlation_pwr_all_lms.pdf")
+                if save_path
+                else None
+            )
+            corr_dict = fit_line_with_ci(
+                fold_errors,
+                fold_uncertainty_values,
+                quantile_thresholds,
+                cmaps,
+                pixel_to_mm=pixel_to_mm_scale,
+                save_path=save_path_fig,
+                to_log=to_log,
+            )
+
+            correlation_dict[model][uncertainty_type] = corr_dict
+
+    return correlation_dict
+
+
+def generate_summary_df(
+    results_dictionary: dict, cols_to_save: list, sheet_name: str, save_location: str
+) -> pd.DataFrame:
+    """
+    Generates pandas dataframe with summary statistics.
+
+    Args:
+        results_dictionary (dict): A dictionary containing results for each quantile bin and uncertainty method.
+            The keys are strings indicating the name of each uncertainty method.
+            The values are dictionaries containing results for each quantile bin.
+        cols_to_save (list): A list of 2-element lists, each containing a string indicating the key in the
+            results_dictionary and a string indicating the name to use for that column in the output dataframe.
+        sheet_name (str): The name of the sheet to create in the output Excel file.
+        save_location (str): The file path and name to use for the output Excel file.
+
+    Returns:
+        pd.DataFrame: A dataframe with statistics including mean error, std error of All and individual
+        landmarks. Also includes the SDR detection rates.
+        The dataframe should have the following structure:
+            df = {
+                "All um col_save_name Mean": value,
+                "All um col_save_name Std": value,
+                "B1 um col_save_name Mean": value,
+                "B1 um col_save_name Std": value,
+                ...
+            }
+    """
+
+    # Save data to csv files
```

Comment:
A utility function for file saving should be created.

Reply:
e.g. wrap the following code in a function like:

def save_dict_to_xlsx(save_path, data):
   if from_dict:
        data = pd.DataFrame.from_dict(data, orient="index")

    with pd.ExcelWriter(save_location, engine="xlsxwriter") as writer:  # pylint: disable=abstract-class-instantiated
        for n, df in (data).items():
            df.to_excel(writer, sheet_name=sheet_name)

Reply:
Have you captured this in the todo list in discussion?